controllers/networkplugins_controller.go

```diff
@@ -17,35 +17,480 @@ limitations under the License.
 package controllers
 
 import (
+	"bytes"
 	"context"
-
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"reflect"
+	"text/template"
+	//"os"
+	//"path/filepath"
 	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/yaml"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	plumberv1 "github.com/platform9/luigi/api/v1"
+	"github.com/platform9/luigi/apply"
+)
+
+const (
+	DEFAULT_NAMESPACE        = "kube-system"
+	MULTUS_IMAGE             = "nfvpe/multus:v3.6"
+	WHEREABOUTS_IMAGE        = "dougbtv/whereabouts:latest"
+	SRIOV_CNI_IMAGE          = "nfvpe/sriov-cni"
+	SRIOV_DP_IMAGE           = "nfvpe/sriov-device-plugin:v3.2"
+	CONFIG_MAP_NAME          = "pf9-networkplugins-config"
+	TEMPLATE_DIR             = "/etc/plugin_templates/"
+	APPLY_DIR                = TEMPLATE_DIR + "apply/"
+	DELETE_DIR               = TEMPLATE_DIR + "delete/"
+	NETWORKPLUGINS_CONFIGMAP = "pf9-networkplugins-configmap"
 )
 
 // NetworkPluginsReconciler reconciles a NetworkPlugins object
 type NetworkPluginsReconciler struct {
 	client.Client
-	Log    logr.Logger
-	Scheme *runtime.Scheme
+	Log         logr.Logger
+	Scheme      *runtime.Scheme
+	currentSpec *plumberv1.NetworkPluginsSpec
+	prevSpec    *plumberv1.NetworkPluginsSpec
+}
+
+type MultusT plumberv1.Multus
+type WhereaboutsT plumberv1.Whereabouts
+type SriovT plumberv1.Sriov
+
+type ApplyPlugin interface {
+	ParseTemplate(string) error
+	ApplyTemplate(string) error
+}
+
+func createFile(config map[string]interface{}, t *template.Template, filename string) error {
+	f, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	err = t.Execute(f, config)
+	if err != nil {
+		fmt.Printf("template.Execute failed: %s\n", err)
+		f.Close()
+		return err
+	}
+	f.Close()
+	return nil
+}
+
+func (multusConfig *MultusT) ParseTemplate(outputDir string) error {
+	config := make(map[string]interface{})
+	if multusConfig.Namespace != "" {
+		config["Namespace"] = multusConfig.Namespace
+	} else {
+		config["Namespace"] = DEFAULT_NAMESPACE
+	}
+
+	if multusConfig.MultusImage != "" {
+		config["MultusImage"] = multusConfig.MultusImage
+	} else {
+		config["MultusImage"] = MULTUS_IMAGE
+	}
+
+	t, err := template.ParseFiles(TEMPLATE_DIR + "multus/multus.yaml")
+	if err != nil {
+		fmt.Printf("ERROR PARSEFILE: %v\n", err)
+		return err
+	}
+
+	if err := createFile(config, t, outputDir+"multus.yaml"); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (multusConfig *MultusT) ApplyTemplate(outputDir string) error {
+	fmt.Println("Applying Multus")
+	return nil
+}
+
+func (whereaboutsConfig *WhereaboutsT) ParseTemplate(outputDir string) error {
```

Comment:
The function name does not really reflect what this function does. Not completely sure what it does. It seems to read and parse files, and then write files.